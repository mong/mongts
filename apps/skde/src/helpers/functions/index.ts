export * from "./imgLoader";
export * from "./filterorderindid";
