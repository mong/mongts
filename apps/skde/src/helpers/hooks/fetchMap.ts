import { useQuery } from "@tanstack/react-query";

export const FetchMap = (filename: string) => {
  const fetching = async (filename) => {
    const response = await fetch(filename);
    if (!response.ok) {
      throw new Error(response.statusText);
    }
    return await response.json();
  };
  return useQuery([`map_${filename}`], () => fetching(filename), {
    staleTime: 1000 * 60 * 60 * 24 * 365,
    refetchOnWindowFocus: false,
    cacheTime: 1000 * 60 * 60 * 24 * 365,
  });
};
