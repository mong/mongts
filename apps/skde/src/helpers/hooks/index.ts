export * from "./useintersectionobserver";
export * from "./useisomorphiclayouteffect";
export * from "./useKeys";
export * from "./useOnCLickOutside";
