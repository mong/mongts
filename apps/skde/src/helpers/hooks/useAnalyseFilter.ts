import { useRouter } from "next/router";

export const useAnalyseFilter = (): [Set<string>, (tag: string) => void] => {
  const router = useRouter();
  const tags = [router.query.tags].flat().filter(Boolean);

  function toggleTag(tag: string) {
    router.replace(
      {
        query: {
          ...router.query,
          tags: tags.includes(tag)
            ? tags.filter((d) => d != tag)
            : tags.concat(tag),
        },
      },
      undefined,
      { shallow: true },
    );
  }
  return [new Set(tags), toggleTag];
};
