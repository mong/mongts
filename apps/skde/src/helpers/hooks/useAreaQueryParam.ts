import { useRouter } from "next/router";

const borhf = {
  "Helse Sør-Øst": [
    "Akershus",
    "OUS",
    "Vestfold",
    "Vestre Viken",
    "Sørlandet",
    "Telemark",
    "Innlandet",
    "Østfold",
    "Diakonhjemmet",
    "Lovisenberg",
  ],
  "Helse Vest": ["Førde", "Bergen", "Fonna", "Stavanger"],
  "Helse Midt-Norge": ["St. Olav", "Møre og Romsdal", "Nord-Trøndelag"],
  "Helse Nord": ["Finnmark", "UNN", "Nordland", "Helgeland"],
};
const borhfSet = new Set(Object.keys(borhf));

/**
 * Returns a hook that toggles the given area in the query param `area` in the current router location.
 * The hook returns a tuple where the first element is the current set of selected areas and
 * the second element is a function to toggle a given area.
 *
 * The hook also handles the special case of the borhf parameter, which is a set of area
 * that are grouped together. If a borhf is selected, all the area in the group are selected.
 * If a area is deselected, all the area in the group are deselected.
 *
 * The hook also handles the special case of the national parameter, which is a area that is
 * not in any group. If the national parameter is set, the hook will not toggle any other area.
 *
 * The hook uses the `useRouter` hook from next/router to get the current router location.
 * The hook uses the `replace` method of the router to update the query param in the current
 * location.
 *
 * The hook is shallow, meaning it will not trigger a full page reload when the query param is
 * updated.
 *
 * @param {string} [national] - The national parameter, which is a area that is not in any group.
 * @returns {[Set<string>, (area: string) => void]} - A tuple where the first element is the current
 * set of selected areas and the second element is a function to toggle a given area.
 */
export const useAreaQueryParam = (
  national: string = "",
  type: string = "area",
): [Set<string>, (area: string) => void] => {
  const router = useRouter();
  const selectedAreas = [router.query[type]].flat().filter(Boolean);

  function toggleArea(area: string) {
    if (area && area != national) {
      const newSelection = new Set(selectedAreas);

      if (newSelection.has(area)) {
        newSelection.delete(area);
        if (borhfSet.has(area)) {
          for (const b of borhf[area]) {
            newSelection.delete(b);
          }
        }
      } else {
        newSelection.add(area);
        if (borhfSet.has(area)) {
          for (const b of borhf[area]) {
            newSelection.add(b);
          }
        }
      }

      for (const [rhf, bohfs] of Object.entries(borhf)) {
        newSelection.delete(rhf);
        if (bohfs.every((area) => newSelection.has(area)))
          newSelection.add(rhf);
      }

      router.replace(
        {
          query: {
            ...router.query,
            [type]: Array.from(newSelection),
          },
        },
        undefined,
        { shallow: true },
      );
    }
  }
  return [new Set(selectedAreas), toggleArea];
};
