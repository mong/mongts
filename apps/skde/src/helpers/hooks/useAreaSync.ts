import { useEffect } from "react";
import { ArrayParam, useQueryParam } from "use-query-params";
import usePostMessageHandler from "./usePostMessageHandler";
import useSiblingFrames from "./useSiblingFrames";

/**
 * A hook that synchronizes the query parameter "area" across sibling
 * windows. It sends out the current value of "area" to other windows and
 * updates the local value when receiving a message from another window.
 *
 * @returns An array where the first element is the current value of "area"
 *          and the second element is a setter function to update the value.
 */
export function useAreaSync() {
  const { siblingFrames, domain } = useSiblingFrames();
  const [message, sendMessage] = usePostMessageHandler(siblingFrames, domain);
  const [areas, setAreas] = useQueryParam("area", ArrayParam);

  useEffect(() => {
    sendMessage({ type: "areas", data: areas });
  }, [areas, sendMessage]);

  useEffect(() => {
    if (message && message.type === "areas") {
      setAreas(message.data);
    }
  }, [message, setAreas]);

  return { areas, setAreas };
}
