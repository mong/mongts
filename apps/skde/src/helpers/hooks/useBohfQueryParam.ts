import { useRouter } from "next/router";

const borhf = {
  "Helse Sør-Øst": [
    "Akershus",
    "OUS",
    "Vestfold",
    "Vestre Viken",
    "Sørlandet",
    "Telemark",
    "Innlandet",
    "Østfold",
    "Diakonhjemmet",
    "Lovisenberg",
  ],
  "Helse Vest": ["Førde", "Bergen", "Fonna", "Stavanger"],
  "Helse Midt-Norge": ["St. Olav", "Møre og Romsdal", "Nord-Trøndelag"],
  "Helse Nord": ["Finnmark", "UNN", "Nordland", "Helgeland"],
};
const borhfSet = new Set(Object.keys(borhf));

export const useBohfQueryParam = (
  national: string = "",
): [Set<string>, (bohf: string) => void] => {
  const router = useRouter();
  const selectedBohfs = [router.query.bohf].flat().filter(Boolean);

  function toggleBohf(bohf: string) {
    if (bohf && bohf != national) {
      const newSelection = new Set(selectedBohfs);

      if (newSelection.has(bohf)) {
        newSelection.delete(bohf);
        if (borhfSet.has(bohf)) {
          for (const b of borhf[bohf]) {
            newSelection.delete(b);
          }
        }
      } else {
        newSelection.add(bohf);
        if (borhfSet.has(bohf)) {
          for (const b of borhf[bohf]) {
            newSelection.add(b);
          }
        }
      }

      for (const [rhf, bohfs] of Object.entries(borhf)) {
        newSelection.delete(rhf);
        if (bohfs.every((bohf) => newSelection.has(bohf)))
          newSelection.add(rhf);
      }

      router.replace(
        {
          query: {
            ...router.query,
            bohf: Array.from(newSelection),
          },
        },
        undefined,
        { shallow: true },
      );
    }
  }
  return [new Set(selectedBohfs), toggleBohf];
};
