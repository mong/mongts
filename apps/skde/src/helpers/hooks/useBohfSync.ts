import { useEffect } from "react";
import { ArrayParam, useQueryParam } from "use-query-params";
import usePostMessageHandler from "./usePostMessageHandler";
import useSiblingFrames from "./useSiblingFrames";

export function useBohfSync() {
  const { siblingFrames, domain } = useSiblingFrames();
  const [message, sendMessage] = usePostMessageHandler(siblingFrames, domain);
  const [bohfs, setBohfs] = useQueryParam("bohf", ArrayParam);

  useEffect(() => {
    sendMessage({ type: "bohfs", data: bohfs });
  }, [bohfs, sendMessage]);

  useEffect(() => {
    if (message && message.type === "bohfs") {
      setBohfs(message.data);
    }
  }, [message, setBohfs]);

  return { bohfs, setBohfs };
}
