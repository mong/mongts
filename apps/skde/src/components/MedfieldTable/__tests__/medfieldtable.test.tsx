import React from "react";
import { MedfieldTable, createMedfieldTableData } from "..";
import { medfieldTableData } from "../../../../test/test_data/data";
import { vi, test, expect } from "vitest";
import * as hooks from "../../../../../../packages/qmongjs/src/helpers/hooks";
import { clockTick } from "qmongjs/src/test/clockTick";
import { render } from "@testing-library/react";

Object.defineProperty(global, "performance", {
  writable: true,
});

const data = createMedfieldTableData(medfieldTableData);

test("Levels counts are correct", () => {
  expect(data).toMatchSnapshot();
});

test("Table renders correctly", async () => {
  vi.spyOn(hooks, "useIndicatorQuery").mockReturnValue({
    data: medfieldTableData,
    isLoading: false,
    error: false,
  });

  const { container } = render(
    <MedfieldTable
      unitNames={["TromsÃ¸"]}
      treatmentYear={2022}
      context={"caregiver"}
      type={"ind"}
      width={800}
    />,
  );

  expect(container).toMatchSnapshot();
});
