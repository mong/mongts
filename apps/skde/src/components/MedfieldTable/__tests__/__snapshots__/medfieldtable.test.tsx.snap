// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Levels counts are correct 1`] = `
[
  ,
  {
    "green": 1,
    "name": "Hjerte- og karsykdommer",
    "red": 1,
    "registers": [
      ,
      {
        "full_name": "Hjerteregister nummer 1",
        "green": 1,
        "name": "hjertereg1",
        "red": 0,
        "yellow": 0,
      },
      {
        "full_name": "Hjerteregister nummer 2",
        "green": 0,
        "name": "hjertereg2",
        "red": 1,
        "yellow": 0,
      },
    ],
    "yellow": 0,
  },
  ,
  {
    "green": 1,
    "name": "Diabetes",
    "red": 0,
    "registers": [
      ,
      ,
      ,
      {
        "full_name": "Diabetesregister nummer 1",
        "green": 1,
        "name": "diabetesreg1",
        "red": 0,
        "yellow": 1,
      },
    ],
    "yellow": 1,
  },
  ,
  {
    "green": 3,
    "name": "Barn",
    "red": 0,
    "registers": [
      ,
      ,
      ,
      ,
      {
        "full_name": "Barneregister nummer 1",
        "green": 1,
        "name": "barnereg1",
        "red": 0,
        "yellow": 0,
      },
      {
        "full_name": "Barneregister nummer 2",
        "green": 2,
        "name": "barnereg2",
        "red": 0,
        "yellow": 0,
      },
    ],
    "yellow": 0,
  },
]
`;

exports[`Table renders correctly 1`] = `
<div>
  <div
    class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
  >
    <table
      aria-label="collapsible table"
      class="MuiTable-root css-rqglhn-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          />
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
            >
              Fagomr√•de
            </h6>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
            >
              Resultat
            </h6>
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root css-2blr9s-MuiTableRow-root"
          style="cursor: pointer;"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <button
              aria-label="expand row"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="KeyboardArrowDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </td>
          <th
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            scope="row"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Hjerte- og karsykdommer
            </p>
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="color: rgb(88, 165, 92); font-size: 1.2rem;"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8z"
              />
            </svg>
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="color: rgb(216, 81, 64); font-size: 1.2rem;"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M464 464H48a16 16 0 0 1-14.07-23.62l208-384a16 16 0 0 1 28.14 0l208 384A16 16 0 0 1 464 464z"
              />
            </svg>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            colspan="6"
            style="padding-bottom: 0px; padding-top: 0px;"
          />
        </tr>
        <tr
          class="MuiTableRow-root css-2blr9s-MuiTableRow-root"
          style="cursor: pointer;"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <button
              aria-label="expand row"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="KeyboardArrowDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </td>
          <th
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            scope="row"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Diabetes
            </p>
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="color: rgb(88, 165, 92); font-size: 1.2rem;"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8z"
              />
            </svg>
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="color: rgb(253, 156, 0); font-size: 1.2rem;"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48z"
              />
            </svg>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            colspan="6"
            style="padding-bottom: 0px; padding-top: 0px;"
          />
        </tr>
        <tr
          class="MuiTableRow-root css-2blr9s-MuiTableRow-root"
          style="cursor: pointer;"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <button
              aria-label="expand row"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="KeyboardArrowDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </td>
          <th
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            scope="row"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Barn
            </p>
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="color: rgb(88, 165, 92); font-size: 1.2rem;"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8z"
              />
            </svg>
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="color: rgb(88, 165, 92); font-size: 1.2rem;"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8z"
              />
            </svg>
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="color: rgb(88, 165, 92); font-size: 1.2rem;"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8z"
              />
            </svg>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            colspan="6"
            style="padding-bottom: 0px; padding-top: 0px;"
          />
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
